cpp/src/arrow/array-test.cc

```diff
@@ -1154,6 +1154,35 @@ TEST_F(TestBinaryBuilder, TestScalarAppend) {
     }
   }
 }
+ 
+TEST_F(TestBinaryBuilder, TestCapacityReserve) {
+  vector<string> strings = {"aaaaa", "bbbbbbbbbb", "ccccccccccccccc", "dddddddddddddddddddd", "eeeeeeeeee"};
+  int N = static_cast<int>(strings.size());
+  int reps = 10;
+  int64_t length = 0;
+  int64_t capacity = 1000;
+  
+
+  ASSERT_OK(builder_->ReserveData(capacity));
+  
+  ASSERT_EQ(length, builder_->value_data_length());
+  ASSERT_EQ(BitUtil::RoundUpToMultipleOf64(capacity), builder_->value_data_capacity());
+ 
+  for (int j = 0; j < reps; ++j) {
+    for (int i = 0; i < N; ++i) {
+      ASSERT_OK(builder_->Append(strings[i]));
+      length += static_cast<int>(strings[i].size());
+   
+      ASSERT_EQ(length, builder_->value_data_length());
+      ASSERT_EQ(BitUtil::RoundUpToMultipleOf64(capacity), builder_->value_data_capacity());
+    }
+  }
```

Comment:
Can you add another call to ReserveData here, like `builder_->ReserveData(500)` to show that the input argument is an incremental amount rather than an absolute amount?